libqtile/core/manager.py

```diff
@@ -444,14 +446,96 @@ def update_net_desktops(self) -> None:
         self.core.update_net_desktops(self.groups, index)
 
     def update_client_list(self) -> None:
-        """Updates the client stack list
+        """Update X11 client list
 
         This is needed for third party tasklists and drag and drop of tabs in
         chrome
         """
         windows = [wid for wid, c in self.windows_map.items() if c.group]
         self.core.update_client_list(windows)
 
+    def update_client_order(self) -> None:
+        """Update X11 client order
+
+        This is needed for third party tasklists and drag and drop of tabs in
+        chrome
+        """
+        windows = [wid for wid, c in self.windows_map.items() if c.group]
+        self.core.update_client_order(windows)
+
+    def change_layer(self, wid, up=True, force=False):
+        """Raise a window above its peers or move it below them, depending on 'up'.
+        Raising a normal window will not lift it above pinned windows etc.
+        """
+        client = self.windows_map.get(wid, None)
+        if client is None:
+            return
+        order = list(self.windows_map.keys())
+        self.windows_map.move_to_end(wid, last=up)
+
+        items = []
+        # we have three flags to check (up, above and below), so there are 8 combinations in total
+        if (up and client.window.net_wm_state_above) or (not up and client.window.net_wm_state_below):  # 2 out of 8
+            # if we're at the top, nobody needs to be raised above us; same in the other direction
+            pass
+        elif up and client.window.net_wm_state_below:  # 2 out of 8 (ignore net_wm_state_above)
+            # if we're at the bottom, almost everybody needs to be put above us
+            items = [(i, w) for i, w in self.windows_map.items() if not w.window.net_wm_state_below]
+        elif not up and client.window.net_wm_state_above:  # 2 out of 8 (ignore net_wm_state_below)
+            # if we're at the top, almost everybody needs to be put below us
+            items = [(i, w) for i, w in self.windows_map.items() if not w.window.net_wm_state_above]
+        elif not client.window.net_wm_state_below and not client.window.net_wm_state_above:  # 2 out of 8
+            # we're just a normal window; make sure we don't interfere with pinned windows
+            items = [
+                (i, w) for i, w in self.windows_map.items()
+                if (w.window.net_wm_state_above if up else w.window.net_wm_state_below)
+            ]
```

Comment:
Any chance you could reformat this a bit? It's quite hard to read, and the lines are very long :)

Reply:
Looks like there might be some repeat calls to `client.window.X` getters that could be prevented by storing the result in a variable

Reply:
Are those expensive? I guess you are talking about the `client.window.net_wm_state`, not the stuff in the list comprehensions, right?

Reply:
They aren't particularly expensive, so it's fine. Just figured that it might also make it less verbose a bit. Perhaps you could do two levels of `if`s i.e. `if up:` and break it down. I know this is being picky, sorry! It will just be helpful in the future when glancing through.

Reply:
If there is no improvement then it's fine as is.

Reply:
I tried to improve it a bit. Pulling the cheks out of the `if` helped quite a bit, I think. Let me know what you think.